File: vibecheck-core/tests/fixtures/lru_cache/claude.js
/**
 * A least-recently-used cache backed by a JavaScript Map.
 *
 * The Map preserves insertion order, which allows us to treat the
 * first entry as the least recently used. On every access or update,
 * the touched entry is deleted and re-inserted at the end of the
 * iteration order, effectively promoting it to most-recently-used.
 */
class LeastRecentlyUsedCache {
  /**
   * Creates a new cache with the given maximum capacity.
   * @param {number} maximumCapacity - The upper bound on stored entries.
   */
  constructor(maximumCapacity) {
    if (maximumCapacity <= 0) {
      throw new RangeError("Cache capacity must be a positive integer");
    }
    this.maximumCapacity = maximumCapacity;
    this.orderedStorage = new Map();
  }

  /**
   * Inserts or updates a key-value pair in the cache.
   *
   * If the key already exists, its value is replaced and the entry is
   * promoted to the most-recently-used position. If the cache is full,
   * the least recently used entry is evicted first. This ensures the
   * cache never grows beyond its configured capacity.
   * @param {*} cacheKey - The key to store.
   * @param {*} cacheValue - The value associated with the key.
   */
  insertEntry(cacheKey, cacheValue) {
    if (this.orderedStorage.has(cacheKey)) {
      this.orderedStorage.delete(cacheKey);
      this.orderedStorage.set(cacheKey, cacheValue);
      return;
    }

    if (this.orderedStorage.size >= this.maximumCapacity) {
      this.evictLeastRecentlyUsedEntry();
    }

    this.orderedStorage.set(cacheKey, cacheValue);
  }

  /**
   * Retrieves the value for the given key, or undefined if absent.
   *
   * Accessing an entry promotes it to the most-recently-used position,
   * which protects it from near-term eviction.
   * @param {*} cacheKey - The key to look up.
   * @returns {*} The cached value, or undefined if not found.
   */
  retrieveEntry(cacheKey) {
    if (!this.orderedStorage.has(cacheKey)) {
      return undefined;
    }

    const cachedValue = this.orderedStorage.get(cacheKey);
    // Note that we must delete and re-insert to move the entry to the
    // end of the Map's iteration order, which represents most-recent.
    this.orderedStorage.delete(cacheKey);
    this.orderedStorage.set(cacheKey, cachedValue);
    return cachedValue;
  }

  /**
   * Returns the number of entries currently stored in the cache.
   * @returns {number} The current entry count.
   */
  currentEntryCount() {
    return this.orderedStorage.size;
  }

  /**
   * Removes the least recently used entry from the cache.
   *
   * The first key in the Map's iteration order is the oldest entry,
   * since every access moves its key to the end via delete-and-reinsert.
   */
  evictLeastRecentlyUsedEntry() {
    const leastRecentKey = this.orderedStorage.keys().next().value;
    this.orderedStorage.delete(leastRecentKey);
  }
}

// Verification that the cache behaves correctly
const demonstrationCache = new LeastRecentlyUsedCache(2);
demonstrationCache.insertEntry("first_key", 100);
demonstrationCache.insertEntry("second_key", 200);

// This access promotes "first_key" to most-recently-used
demonstrationCache.retrieveEntry("first_key");

demonstrationCache.insertEntry("third_key", 300);

// Note that "second_key" was evicted because it was the least recently used
console.assert(demonstrationCache.retrieveEntry("second_key") === undefined);
console.assert(demonstrationCache.retrieveEntry("first_key") === 100);
console.assert(demonstrationCache.currentEntryCount() === 2);

module.exports = { LeastRecentlyUsedCache };

File: vibecheck-core/tests/fixtures/lru_cache/claude.rs
use std::collections::{HashMap, VecDeque};

/// A least-recently-used cache with a fixed maximum capacity.
///
/// This implementation uses a `HashMap` for constant-time key lookups
/// and a `VecDeque` to track access recency. When the cache exceeds
/// its maximum capacity, the least recently accessed entry is evicted.
pub struct LeastRecentlyUsedCache<KeyType, ValueType>
where
    KeyType: Eq + std::hash::Hash + Clone,
{
    stored_entries: HashMap<KeyType, ValueType>,
    access_order_tracker: VecDeque<KeyType>,
    maximum_capacity: usize,
}

impl<KeyType, ValueType> LeastRecentlyUsedCache<KeyType, ValueType>
where
    KeyType: Eq + std::hash::Hash + Clone,
{
    /// Creates a new cache that holds at most `maximum_capacity` entries.
    ///
    /// Note that a capacity of zero is allowed but means every insertion
    /// will immediately evict the entry.
    pub fn with_capacity(maximum_capacity: usize) -> Self {
        Self {
            stored_entries: HashMap::with_capacity(maximum_capacity),
            access_order_tracker: VecDeque::with_capacity(maximum_capacity),
            maximum_capacity,
        }
    }

    /// Inserts a key-value pair into the cache.
    ///
    /// If the key already exists, its value is updated and it is promoted
    /// to the most-recently-used position. If inserting a new key would
    /// exceed the maximum capacity, the least recently used entry is
    /// evicted first. This ensures the cache never grows beyond its bound.
    pub fn insert_entry(&mut self, cache_key: KeyType, cache_value: ValueType) {
        if self.stored_entries.contains_key(&cache_key) {
            self.promote_key_to_most_recent(&cache_key);
            self.stored_entries.insert(cache_key, cache_value);
            return;
        }

        if self.stored_entries.len() >= self.maximum_capacity {
            self.evict_least_recently_used_entry();
        }

        self.access_order_tracker.push_back(cache_key.clone());
        self.stored_entries.insert(cache_key, cache_value);
    }

    /// Retrieves a reference to the value associated with the given key.
    ///
    /// Accessing an entry promotes it to the most-recently-used position,
    /// which protects it from near-term eviction.
    pub fn retrieve_entry(&mut self, cache_key: &KeyType) -> Option<&ValueType> {
        if self.stored_entries.contains_key(cache_key) {
            self.promote_key_to_most_recent(cache_key);
            return self.stored_entries.get(cache_key);
        }
        None
    }

    /// Returns the number of entries currently stored in the cache.
    pub fn current_entry_count(&self) -> usize {
        self.stored_entries.len()
    }

    /// Moves the given key to the back of the access-order tracker,
    /// marking it as the most recently used entry. This ensures that
    /// frequently accessed entries survive eviction longer.
    fn promote_key_to_most_recent(&mut self, cache_key: &KeyType) {
        self.access_order_tracker
            .retain(|existing_key| existing_key != cache_key);
        self.access_order_tracker.push_back(cache_key.clone());
    }

    /// Removes the least recently used entry from both the access-order
    /// tracker and the underlying storage map.
    fn evict_least_recently_used_entry(&mut self) {
        if let Some(evicted_key) = self.access_order_tracker.pop_front() {
            self.stored_entries.remove(&evicted_key);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn evicts_least_recently_used_when_capacity_exceeded() {
        let mut cache = LeastRecentlyUsedCache::with_capacity(2);
        cache.insert_entry("first_key", 100);
        cache.insert_entry("second_key", 200);
        cache.insert_entry("third_key", 300);

        // Note that "first_key" was evicted because it was the oldest entry
        assert!(cache.retrieve_entry(&"first_key").is_none());
        assert_eq!(cache.retrieve_entry(&"second_key"), Some(&200));
        assert_eq!(cache.retrieve_entry(&"third_key"), Some(&300));
    }
}
